### PR TITLE
crimson/net: misc fixes and features for crimson-messenger tests

### PR DESCRIPTION
**Pull request by cyx1231st**

### CODE REVIEW

File: src/crimson/net/SocketMessenger.h [hunk 1]
```diff
@@ -43,39 +45,74 @@ class SocketMessenger final : public Messenger {
...
+  seastar::future<> do_start(Dispatcher *disp);
+  seastar::foreign_ptr<ConnectionRef> do_connect(const entity_addr_t& peer_addr,
+                                                 const entity_type_t& peer_type);
+  seastar::future<> do_shutdown();
+  // conn sharding options:
+  // 0. Compatible (master_sid >= 0): place all connections to one master shard
+  // 1. Simplest (master_sid < 0): sharded by ip only
+  // 2. Balanced (not implemented): sharded by ip + port + nonce,
+  //        but, need to move SocketConnection between cores.
+  seastar::shard_id locate_shard(const entity_addr_t& addr);
```
Comment (tchaikov) [inline src/crimson/net/SocketMessenger.h]: i'd suggest something like:
```c++
namespace ceph::net::placement_policy {
  class Pinned {
    const seastar::shard_id where;
    Pinned(seastar::shard_id shard_id)
      : where{shard_id}
    {}
    seasetar::shard_id locate(const entity_addr_t&) const {
      return where;
    }
  };
  class ShardByAddr {
    seasetar::shard_id locate(const entity_addr_t& addr) const {
      std::size_t seed = 0;
      boost::hash_combine(seed, addr.u.sin.sin_addr.s_addr);
      //boost::hash_combine(seed, addr.u.sin.sin_port);
      //boost::hash_combine(seed, addr.nonce);
      return seed % seastar::smp::count;
    }
  };
}

template <class PlacementPolicy=placement_policy::Pinned>
SocketMessenger(const entity_name_t& myname,
                const std::string& logic_name,
                uint32_t nonce,
                const PlacementPolicy& placement)
  : placement{placement}
{}
  seastar::future<ceph::net::ConnectionXRef> connect(const entity_addr_t& peer_addr,
                                                     const entity_type_t& peer_type) {
    return container().invoke_on(placement.locate(peer_addr),
                                 [...] (...) {
      // ...
    });
  }                          
};
```

File: src/crimson/net/SocketConnection.cc [hunk 2]
```diff
@@ -276,12 +276,13 @@ seastar::future<> SocketConnection::write_message(MessageRef msg)
...
   }
   // write as a seastar::net::packet
-  return socket->write_flush(std::move(bl))
-    .then([this, msg = std::move(msg)] {
-      if (!policy.lossy) {
-        sent.push(std::move(msg));
-      }
-    });
+  return socket->write_flush(std::move(bl));
+  // TODO: lossless policy
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/net/SocketConnection.cc]: As far as I know, the following are not implemented correctly for lossless policy:
- Replacing logic is wrong and partially implemented.
- Wait and standby connection states not implemented.
- Fault handling in all connection states.
- Message acknowledgement, requeue and correction.
- Connection racing.

I assume lossless will be implemented with v2 protocol.

File: src/crimson/net/SocketConnection.cc [hunk 3]
```diff
@@ -61,37 +62,50 @@ SocketConnection::get_messenger() const {
...
 seastar::future<> SocketConnection::send(MessageRef msg)
 {
-  if (state == state_t::closing)
-    return seastar::now();
-  return seastar::with_gate(pending_dispatch, [this, msg=std::move(msg)] {
-      return do_send(std::move(msg))
-        .handle_exception([this] (std::exception_ptr eptr) {
-          logger().warn("{} send fault: {}", *this, eptr);
-          close();
+  return seastar::smp::submit_to(shard_id(), [this, msg=std::move(msg)] {
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/crimson/net/SocketConnection.cc]: i see. so w.r.t. m:n mapping, we have to do perform the i/o on some **certain** core, otherwise different cores talking to the same peer OSD could racing with each other.

File: src/test/crimson/perf_crimson_msgr.cc [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+                    if (stop) {
+                      logger().info("{}: finished sending {} OSDOPs with {} keepalives",
+                                    *conn, count_ping, count_keepalive);
+                    }
+                    return stop;
+                  },
+                  [this, conn, &count_ping, &count_keepalive] {
+                    return seastar::repeat([this, conn, &count_ping, &count_keepalive] {
+                        if (keepalive_dist(rng)) {
+                          return conn->keepalive()
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/test/crimson/perf_crimson_msgr.cc]: the `conn->keepalive()`'s counterpart is the new extracted `Client::send_msg()`, maybe no need to extract this?

File: src/test/crimson/perf_crimson_msgr.cc [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+      ceph::net::create_sharded<test_state::Server>(),
+      ceph::net::create_sharded<test_state::Client>(rounds, keepalive_ratio, bs, depth))
+    .then([rounds, keepalive_ratio, addr, mode](result_tuple t) {
+      test_state::Server& server = std::get<0>(t).get0();
+      test_state::Client& client = std::get<1>(t).get0();
+      // start servers and clients
+      entity_addr_t target_addr;
+      target_addr.parse(addr.c_str(), nullptr);
+      target_addr.set_type(entity_addr_t::TYPE_LEGACY);
+      if (mode == 0) {
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/test/crimson/perf_crimson_msgr.cc]: fixed.

File: src/crimson/net/SocketMessenger.h (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/net/SocketMessenger.h]: will do this in another PR.

File: src/crimson/osd/main.cc [hunk 6]
```diff
@@ -63,8 +64,18 @@ int main(int argc, char* argv[])
...
       }).then([&] {
-        return osd.start_single(std::stoi(local_conf()->name.get_id()),
-                                static_cast<uint32_t>(getpid()));
+        int id = std::stoi(local_conf()->name.get_id());
+        auto nonce = static_cast<uint32_t>(getpid());
+        return seastar::when_all_succeed(
+            ceph::net::Messenger::create(entity_name_t::OSD(id), "cluster", nonce, 0),
+            ceph::net::Messenger::create(entity_name_t::OSD(id), "public", nonce, 0),
+            ceph::net::Messenger::create(entity_name_t::OSD(id), "hb_front", nonce, 0),
+            ceph::net::Messenger::create(entity_name_t::OSD(id), "hb_back", nonce, 0)
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/osd/main.cc]: > why can't we create messengers where the they are used?

I agree. Actually we can, and we should create messengers where they are used in the 1:1 mapping prototype, so that a compatible-mode messenger can be correctly created, while, be limited within the core where the OSD/Heartbeat/Monc initiated. Please refer to the updated code.

> i'd prefer managing the life cycle of Messenger explicitly by its consumer by calling start() and stop(), and only use ceph::net::Messenger::create() in the unittest and demo.

The purpose of `ceph::net::Messenger::create()` is to allocate `Messenger` in all available cores, allow it to listen and accept from all cores after started, so it won't stop working if reactors > 1. It just make our 1:1 mapping prototype more robust, regardless whether we decide to run OSD instance as shards, or to introduce another reactor to offload some tasks. `start()` and `shutdown()` are still used to let consumer manage the lifecycle of its Messenger. The only difference is we can no longer allocate Messenger inside the constructor of OSD/Heartbeat, because `ceph::net::Messenger::create()` returns a future now.

> as the gracefully shutdown would be a nightmare if we leave it to the create() wrapper.

That's why I decouple `shutdown()` with `stop()`, it allows consumer to end the lifecycle of its Messenger explicitly and gracefully. And the final destruction will be delayed until the engine exits, not with the destruction of the consumer.

File: src/test/crimson/test_messenger.cc [hunk 7]
```diff
@@ -33,6 +34,7 @@ static seastar::future<> test_echo(unsigned rounds,
         : public ceph::net::Dispatcher,
           public seastar::peering_sharded_service<Server> {
       ceph::net::Messenger *msgr = nullptr;
+      MessageRef msg_pong = MessageRef{new MPing(), false};
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/test/crimson/test_messenger.cc]: fixed.

File: src/crimson/osd/heartbeat.cc [hunk 8]
```diff
@@ -48,17 +46,20 @@ seastar::future<> Heartbeat::start(entity_addrvec_t front_addrs,
...
+  return seastar::when_all_succeed(
+      front_msgr->try_bind(front_addrs,
+                           local_conf()->ms_bind_port_min,
+                           local_conf()->ms_bind_port_max),
+      back_msgr->try_bind(back_addrs,
+                          local_conf()->ms_bind_port_min,
+                          local_conf()->ms_bind_port_max))
+    .then([this] {
+      return seastar::when_all_succeed(
+        front_msgr->start(this), back_msgr->start(this));
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/osd/heartbeat.cc]: fixed.

File: src/crimson/net/Dispatcher.h [hunk 9]
```diff
@@ -54,6 +55,11 @@ class Dispatcher {
   }
   virtual seastar::future<std::unique_ptr<AuthAuthorizer>>
   ms_get_authorizer(peer_type_t);
+
+  // get the locol dispatcher shard if it is accessed by another core
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/net/Dispatcher.h]: fixed.

File: src/crimson/net/Messenger.h [hunk 10]
```diff
@@ -17,11 +17,16 @@
...
 #include "Fwd.h"
+#include "crimson/thread/Throttle.h"
+#include "msg/Policy.h"
 
 class AuthAuthorizer;
 
 namespace ceph::net {
 
+using SocketPolicy = ceph::net::Policy<ceph::thread::Throttle>;
+using Throttle = ceph::thread::Throttle;
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/net/Messenger.h]: fixed.

File: src/crimson/net/Messenger.h [hunk 11]
```diff
@@ -89,6 +94,12 @@ class Messenger {
 
   virtual void print(ostream& out) const = 0;
 
+  virtual void set_default_policy(const SocketPolicy& p) = 0;
+
+  virtual void set_policy(entity_type_t peer_type, const SocketPolicy& p) = 0;
+
+  virtual void set_policy_throttler(entity_type_t peer_type, Throttle* throttle) = 0;
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/net/Messenger.h]: splitted in 4a1053bee7b2b4ad0bdec01a77319df70e7c7c12

File: src/test/crimson/test_messenger.cc [hunk 12]
```diff
@@ -19,175 +29,339 @@ static seastar::future<> test_echo(unsigned rounds,
...
+      return seastar::when_all_succeed(
+          server1->init(entity_name_t::OSD(0), "server1", 1, addr1),
+          server2->init(entity_name_t::OSD(1), "server2", 2, addr2),
+          client1->init(entity_name_t::OSD(2), "client1", 3),
+          client2->init(entity_name_t::OSD(3), "client2", 4))
+      // dispatch pingpoing
+        .then([client1, client2, server1, server2] {
+          return seastar::when_all_succeed(
+              client1->dispatch_pingpong(server1->msgr->get_myaddr(), true),
+              client1->dispatch_pingpong(server2->msgr->get_myaddr(), false),
```
Comment (tchaikov) [inline src/test/crimson/test_messenger.cc]: why do we need two tests for testing `!foreign_dispatch`:
```c++
client1->dispatch_pingpong(server2->msgr->get_myaddr(), false),
```
and 
```c++
client2->dispatch_pingpong(server1->msgr->get_myaddr(), false),
```
?
simply to see if a `SocketConnection` is able to handle i/o sent from different shards in parallel?

File: src/test/crimson/test_messenger.cc [hunk 13]
```diff
@@ -198,19 +372,19 @@ int main(int argc, char** argv)
...
-    return test_echo(rounds, keepalive_ratio)
-    .then([] {
-      return test_concurrent_dispatch();
-    }).then([] {
-      std::cout << "All tests succeeded" << std::endl;
-    }).handle_exception([] (auto eptr) {
-      std::cout << "Test failure" << std::endl;
-      return seastar::make_exception_future<>(eptr);
+  return app.run(argc, argv, [&app] {
+      auto&& config = app.configuration();
```
Comment (tchaikov) [inline src/test/crimson/test_messenger.cc]: i'd suggest avoid the massive reindent unless we are changing the indented code at the same time.

File: src/crimson/osd/main.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/crimson/osd/main.cc]: i see. the life cycle of a shared service is managed by its container, namely, `seastar::sharded<>` and friends. and `seastar::shared<>` does not allow us to update an `seastar::sharded<>` with  a service instance created as a member variable.

what we can is but to promote these member variables to "global" variables even if they are only used by a single class, which does not have to be sharded. but if it is, the `Messenger` should be sharded as well.

File: src/test/crimson/test_messenger.cc (hunk 12 above)
Comment (cyx1231st) [reply src/test/crimson/test_messenger.cc]: Yes, the four `dispatch_pingpong(...)` combinations are to test connecting in parallel, accepting in parallel, and operating the connection reference from a foreign/local core.

File: src/crimson/net/Dispatcher.h [hunk 14]
```diff
@@ -54,6 +55,11 @@ class Dispatcher {
   }
   virtual seastar::future<std::unique_ptr<AuthAuthorizer>>
   ms_get_authorizer(peer_type_t);
+
+  // get the local dispatcher shard if it is accessed by another core
+  virtual Dispatcher* get_local_shard() {
```
Comment (tchaikov) [inline src/crimson/net/Dispatcher.h]: i have a feeling that we are going to have a melange of the concept of `shared<>` and `Service` here. as you are implying that `Service` should be aware that it's being sharded. the caller site of this method looks like
```c++
seastar::future<> SocketMessenger::start(Dispatcher *disp) {
   return container().invoke_on_all([disp](auto& msgr) {
       return msgr.do_start(disp->get_local_shard());
     });
}
```
but i think the caller of `SocketMessenger::start()` should be responsible for passing the *local* service to to *local* instance of `SocketMessenger`.

File: src/test/crimson/test_messenger.cc (hunk 13 above)
Comment (cyx1231st) [reply src/test/crimson/test_messenger.cc]: Sorry, didn't realize that this part doesn't change much. `test_messenger.cc` was originally rewritten to verify the sharded-messenger prototype, then supplemented with the original test cases.

File: src/crimson/net/Dispatcher.h (hunk 14 above)
Comment (cyx1231st) [reply src/crimson/net/Dispatcher.h]: The major purpose of `get_local_shard()` is to allow any caller holding an abstract `Dispatcher` to get the local shard if the concrete `Dispatcher` class is sharded. While, if the the concrete `Dispatcher` class is not sharded, it can simply reuse the default implementation of `Dispatcher::get_local_shard()` to get the compatible behavior.

The specific example here, is to let `SocketMessenger` to always get a valid pointer regardless of whether the concrete Dispatcher is sharded or not.

File: src/test/crimson/test_messenger.cc (hunk 12 above)
Comment (tchaikov) [reply src/test/crimson/test_messenger.cc]: could accompany it with a comment to explain what's the intention of having multiple client here?

File: src/crimson/net/Dispatcher.h (hunk 14 above)
Comment (tchaikov) [reply src/crimson/net/Dispatcher.h]: well, thought there was some alternatives to gap the difference. but let's keep it this way. will come up with another (hopefully better) solution later on.

**Comment by tchaikov** [comment]
@cyx1231st could you address the pending comments? then i think it's good to merge. we can continue on based on this.

File: src/crimson/net/Connection.h [hunk 15]
```diff
@@ -15,17 +15,18 @@
...
 
 namespace ceph::net {
 
 using seq_num_t = uint64_t;
+using SharedPtr = seastar::shared_ptr<seastar::shared_ptr_count_base>;
 
-class Connection : public boost::intrusive_ref_counter<Connection,
-						       boost::thread_unsafe_counter> {
+class Connection : public seastar::enable_shared_from_this<Connection> {
+  SharedPtr priv;
```
Comment (tchaikov) [inline src/crimson/net/Connection.h]: i don't think this change is related to the enabling of x-core communication. could you extract it into its own commit?

File: src/crimson/net/Connection.h [hunk 16]
```diff
@@ -34,12 +35,18 @@ class Connection : public boost::intrusive_ref_counter<Connection,
   Connection() {}
   virtual ~Connection() {}
 
+  void set_priv(const SharedPtr& o) {
+    priv = o;
+  }
+  SharedPtr get_priv() {
```
Comment (tchaikov) [inline src/crimson/net/Connection.h]: does it make sense to templaterize this method by adding a template parameter, like
```c++
template<class T> auto get_priv() {
  return boost:: dynamic_pointer_cast<T>(priv);
}
```
as it's rare that somebody is really interested in a `SharedPtr`.

File: src/crimson/net/Messenger.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/net/Messenger.h]: fixed. please check

File: src/crimson/net/SocketMessenger.cc [hunk 17]
```diff
@@ -31,59 +33,106 @@ namespace {
...
-    } catch (const std::system_error& e) {
-      logger().debug("{}: try_bind: {} already used", *this, port);
-      if (port == max_port) {
-	throw;
-      }
-    }
-  }
+  ceph_assert(min_port <= max_port);
+  return seastar::do_with(uint32_t(min_port),
+    [this, max_port, addr] (auto &port) {
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/net/SocketMessenger.cc]: fixed.

File: src/test/crimson/test_messenger.cc (hunk 12 above)
Comment (cyx1231st) [reply src/test/crimson/test_messenger.cc]: comment added.

File: src/test/crimson/test_messenger.cc (hunk 13 above)
Comment (cyx1231st) [reply src/test/crimson/test_messenger.cc]: fixed.

File: src/crimson/net/Connection.h (hunk 15 above)
Comment (cyx1231st) [reply src/crimson/net/Connection.h]: splitted to 3cb78640d0fb831f903e6d3f65ba13c0600b9052

File: src/crimson/net/Connection.h (hunk 16 above)
Comment (cyx1231st) [reply src/crimson/net/Connection.h]: Implemented, but not quite confident...

Because `seastar::enable_shared_from_this` privately inherits `seastar::shared_ptr_count_base`, neither `seastar::dynamic_pointer_cast` nor `seastar::static_pointer_cast` works here.

Investigated a lot, the only solution I found is to use `reinterpret_cast` and return a raw pointer of `T` from `priv.get()`

Comment (cyx1231st) [reply src/crimson/net/Connection.h]: A better one might be to return `reinterpret_cast<T*>(priv.get())->shared_from_this()` instead of raw pointer, what do you think?

Comment (tchaikov) [reply src/crimson/net/Connection.h]: discussed over IRC, we will drop the `{set,get}_priv()` change. and use a `std::map<>` for tracking the mapping from connection to session instead.

**Comment by cyx1231st** [comment]
Dropped `{set,get}_priv()`.
